core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java

```diff
@@ -58,7 +58,6 @@ protected HandlerResult doAuthentication(final Credential credential) throws Gen
         }
 
         userPass.setUsername(transformedUsername);
-        userPass.setPassword(this.passwordEncoder.encode(userPass.getPassword()));
```

Comment:
I would agree that most if not all authN handlers should be using password encoders where possible. But per my reckoning, both method calls should be used. The flow would be that:

- Password is encoded first using the encoder, i.e. `encode`.
- Authentication uses the encoded password to find the account and match on it.
- Double check the provided password, if/when necessary, by the handler and make sure `matches()` actually does a match. 

In the case of QueryDatabaseAuthenticationHandler for instance:

```java
        final String username = credential.getUsername();
        final String password = credential.getPassword();
       //password is encoded already by the parent class that handles credential prep.

        try {

            //execute authN. 
            final String dbPassword = getJdbcTemplate().queryForObject(this.sql, String.class, username);

           //this needs to turn into a passwordEncoder.matches()
            if (!StringUtils.equals(password, dbPassword)) {
                throw new FailedLoginException("Password does not match value on record.");
            }
....
```


The only problem as you describe is that BCrypt wants the raw password which we lose if encode it first in the parent class. So you'll just need to massage the calls to keep a non-encoded version of it, and that would keep it consistent across all encoders, etc. Something as simple/dumb as perhaps (without too much complication):

```java
authenticateUsernamePasswordInternal(userPass, originalCredential);
```

Then let the specific authentication handler deal with the `encoder.matches` since now you have both, and that shall work with all encoders since they all expect the raw password.

As for QueryAndEncodeDatabaseAuthenticationHandler, let's discuss and track that in a separate issue.